### PR TITLE
retry openai completion api

### PR DESCRIPTION
**Pull request by wskish**
The OpenAI completion API can fail due to various reasons including network errors or more commonly their RateLimitError, or ServiceUnavailableError.

This adds a retry() decorator that implements an exponential backoff retry strategy and uses that decorator to wrap the openai create() call in the normal non-streaming use case.

The default behavior of the retry decorator is to retry 5 times, with an initial delay of 0.5 seconds with subsequent tries doubling the delay.

### CODE REVIEW

File: langchain/llms/openai.py [hunk 1]
```diff
@@ -155,7 +155,10 @@ def _generate(
         # Includes prompt, completion, and total tokens used.
         _keys = {"completion_tokens", "prompt_tokens", "total_tokens"}
         for _prompts in sub_prompts:
-            response = self.client.create(prompt=_prompts, **params)
+            @retry()
+            def create():
+                return self.client.create(prompt=_prompts, **params)
```
Comment (zhengligs) [inline langchain/llms/openai.py:160]: There are multiple places `client.create` is called.

Comment (wskish) [reply langchain/llms/openai.py:160]: The other usage of client.create() is as a generator for the streaming case; I need to think more about how to wrap the generator for retry.  To make it fully robust to against transient failures that occur mid-stream may require retry logic within the openai client.

File: langchain/utils.py [hunk 2]
```diff
@@ -19,3 +24,46 @@ def get_from_dict_or_env(
             f" `{env_key}` which contains it, or pass"
             f"  `{key}` as a named parameter."
         )
+
+
+
+def retry(ExceptionToCheck: Type[Exception] = Exception, tries: int = 5, delay: float = 0.5, backoff: int = 2, ExceptionToRaise: Type[Exception] =AssertionError) -> Callable:
```
Comment (hwchase17) [inline langchain/utils.py:30]: for variable names, can we do `exception_to_check`, etc?

also, for this specific usage can we pass in the openai rate limit error class

Comment (wskish) [reply langchain/utils.py:30]: Sure happy to rework it to match style. 

For the default openapi exceptions to catch, I regularly see RateLimitError and ServiceUnavailableError. This catches those by default along with other random network issues.   Do you want to limit the retry to a certain class of errors, or exclude certain errors?  

Also today I started running into weird RateLimitErrors that required more aggressive backoffs.  (I think openai token rate limiting logic is currently broken.) We may need to increase the default number of tries beyond  5.

File: langchain/utils.py [hunk 3]
```diff
@@ -19,3 +24,51 @@ def get_from_dict_or_env(
             f" `{env_key}` which contains it, or pass"
             f"  `{key}` as a named parameter."
         )
+
+
+def retry(
```
Comment (ret2libc) [inline langchain/utils.py:29]: OpenAI documentation also suggests using https://github.com/jd/tenacity , though maybe it is too big for only few things needed.

**Comment by agola11** [comment]
Hey all -- put up a PR to address this and used tenacity: https://github.com/hwchase17/langchain/pull/849

**Comment by hwchase17** [comment]
thanks for kicking this off @wskish and thanks for updating it use tenacity @agola11 ! gonna merge ankush's in and close this one

**Pull request closed without merging**